File: 653. Two Sum IV - Input is a BST/main.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    HashSet<Integer> result = new HashSet<>();
    // dfs
    public boolean findTarget(TreeNode root, int k) {
        if (root==null){
            return false;
        }
        if (result.contains(k-root.val)){
            return true;
        } else {
            result.add(root.val);
            return findTarget(root.left, k) || findTarget(root.right,k);
        }
    }
    
    // bfs
    // public boolean findTarget(TreeNode root, int k) {
    //     if (root== null) return false;
    //     Queue<TreeNode> nodeQ = new LinkedList<TreeNode>();
    //     nodeQ.offer(root);
    //     while(nodeQ.size()!=0){
    //         TreeNode node = nodeQ.poll();
    //         if (result.contains(k-node.val)) return true;
    //         else {
    //             result.add(node.val);
    //             if (node.left != null) nodeQ.offer(node.left);
    //             if (node.right!=null) nodeQ.offer(node.right);
    //         }
    //     }
    //     return false;
    // }
}

File: 653. Two Sum IV - Input is a BST/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findTarget(self, root: TreeNode, k: int) -> bool:
        
        # dfs recursive
#         def dfs(node):
#             if node:
#                 table.add(node.val)
#                 dfs(node.left)
#                 dfs(node.right)
#             else:
#                 return
            
 
#         if not root:
#             return False
        
#         table = set()
#         dfs(root)
#         for num in table:
#             target = k-num
#             if target in table and target != num:
#                 return True
#         return False

        # bfs iterative
        def bfs(node):
            stack = [node]
            for node in stack:
                if (k-node.val) in table:
                    return True
                else:
                    table.add(node.val)
                    if node.left:
                        stack.append(node.left)
                    if node.right:
                        stack.append(node.right)
            return False
        
        if not root:
            return False
        table = set()
        return bfs(root)